src/jit/lowerxarch.cpp

```diff
@@ -1110,6 +1116,23 @@ void Lowering::TreeNodeInfoInitCall(GenTreeCall* call)
         // computed into a register.
         if (!call->IsFastTailCall())
         {
+#ifdef _TARGET_X86_
+            // On x86, we need to generate a very specific pattern for indirect VSD calls:
+            //
+            //    3-byte nop
+            //    call dword ptr [eax]
+            //
+            // Where EAX is also used as an argument to the stub dispatch helper. Make
+            // sure that the call target address is computed into EAX in this case.
+            if (((call->gtFlags & GTF_CALL_VIRT_KIND_MASK) == GTF_CALL_VIRT_STUB) && (call->gtCallType == CT_INDIRECT))
+            {
+                assert(ctrlExpr->isIndir());
+
+                ctrlExpr->gtGetOp1()->gtLsraInfo.setDstCandidates(l, RBM_EAX);
```

Comment:
say ctrlExpr is GT_IND(GT_ADDDR)

GT_IND is getting marked as contained.  So GT_ADDR is getting consumed by GT_CALL node and not by GT_IND.  At the Use of GT_ADDR (which is GT_CALL codegen), we would like it be in EAX.

On the other hand if we set DstCandidates on GT_ADDR, then at the time of producing (i.e. at Def position) the value GT_ADDR needs to be moved to EAX.  But say if GT_ADDR gets spilled after it is produced in EAX and later gets reloaded to a different reg  (say EBX) at the point of use, then at the point of Use we are supposed to use EBX. Using EAX at the point of use would be incorrect.  But the reason it mostly works as required is as i said nothing comes between ctrlExpr and GT_CALL node.

---
In reply to: [79257304](https://github.com/dotnet/coreclr/pull/7232#discussion_r79257304) [](ancestors = 79257304)